src/java.base/share/classes/jdk/internal/lang/stable/StableValueImpl.java

```diff
@@ -0,0 +1,191 @@
+    // | -------------- |  ------------ |
+    // | null           |  Unset        |
+    // | nullSentinel() |  Set(null)    |
+    // | other          |  Set(other)   |
+    //
+    @Stable
+    private volatile Object value;
+
+    // Only allow creation via the factory `StableValueImpl::newInstance`
+    private StableValueImpl() {}
+
+    @ForceInline
+    @Override
+    public boolean trySet(T value) {
+        if (this.value != null) {
+            return false;
+        }
+        // Mutual exclusion is required here as `computeIfUnset` might also
+        // attempt to modify the `wrappedValue`
+        synchronized (this) {
+            return wrapAndCas(value);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public void setOrThrow(T value) {
+        if (!trySet(value)) {
+            throw new IllegalStateException("Cannot set the underlying data to " + value +
+                    " because the underlying data is already set: " + this);
+        }
+    }
+
+    @ForceInline
+    @Override
+    public T orElseThrow() {
+        final Object t = value;
+        if (t == null) {
+            throw new NoSuchElementException("No underlying data set");
+        }
+        return unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public T orElse(T other) {
+        final Object t = value;
+        return (t == null) ? other : unwrap(t);
+    }
+
+    @ForceInline
+    @Override
+    public boolean isSet() {
+        return value != null;
+    }
+
+    @ForceInline
+    @Override
+    public T orElseSet(Supplier<? extends T> supplier) {
+        final Object t = value;
+        return (t == null) ? computeIfUnsetSlowPath(supplier) : unwrap(t);
+    }
+
+    @DontInline
+    private synchronized T computeIfUnsetSlowPath(Supplier<? extends T> supplier) {
+        final Object t = value;
+        if (t == null) {
+            final T newValue = supplier.get();
+            // The mutex is reentrant so we need to check if the value was actually set.
+            return wrapAndCas(newValue) ? newValue : orElseThrow();
+        }
+        return unwrap(t);
+    }
+
+    // The methods equals() and hashCode() should be based on identity (defaults from Object)
+
+    @Override
+    public String toString() {
+        final Object t = value;
+        return t == this
+                ? "(this StableValue)"
+                : "StableValue" + renderWrapped(t);
+    }
+
+    // Internal methods shared with other internal classes
+
+    @ForceInline
+    public Object wrappedValue() {
+        return value;
+    }
+
+    static String renderWrapped(Object t) {
+        return (t == null) ? ".unset" : "[" + unwrap(t) + "]";
+    }
+
+    // Private methods
+
+    @ForceInline
+    private boolean wrapAndCas(Object value) {
+        // This upholds the invariant, a `@Stable` field is written to at most once
+        return UNSAFE.compareAndSetReference(this, UNDERLYING_DATA_OFFSET, null, wrap(value));
```

Comment:
There is no need for a cas here as all setters have to hold the lock. We should have a dedicated private `set` that asserts `Thread.holdsLock(this)`.